### PR TITLE
Add a graph generation template and entry points

### PR DESCRIPTION
**Pull request by jcagme**
This allows all repos to opt-in to add a leg to generate the graph files and upload them to the artifacts folder.

Test build: https://dnceng.visualstudio.com/internal/_build/results?buildId=117649

### CODE REVIEW

**Comment by jcagme** [comment]
@chcosta can you PTAL?

File: eng/common/templates/jobs/jobs.yml [hunk 1]
```diff
@@ -13,6 +13,12 @@ parameters:
 
   # Optional: Enable publishing to the build asset registry
   enablePublishBuildAssets: false
+  
+  # Optional: Enable generating the graph files at the end of the build
+  enableGraphFilesGeneration: false
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: What if the schema looked like...

```YAML
graphFileGeneration:
  enabled: false
  includeToolset: false
```
… then below, the condition looks like...

```YAML
- ${{ if and(eq(parameters.graphFileGeneration.enabled, true), … }}:
  - template: …
```

This means that the caller specifies …

```YAML
template: eng\common\templates\jobs\jobs.yml
parameters:
  enablePublishBuildAssets: true
  enableTelemetry: true
  graphFileGeneration:
    enabled: true
  …
```

My hope is that this makes the experience (from the caller perspective) cleaner and also provides some grouping for these parameters.

Comment (jcagme) [reply eng/common/templates/jobs/jobs.yml]: I wanted to do exactly this but found no example like that in our existing YMLs. Thanks YML man!

**Comment by jcagme** [comment]
@chcosta it looks fine https://dnceng.visualstudio.com/internal/_build/results?buildId=119144

**Comment by chcosta** [comment]
I don't see any changes here in your PR

File: eng/common/templates/jobs/jobs.yml [hunk 2]
```diff
@@ -68,4 +74,22 @@ jobs:
...
       runAsPublic: ${{ parameters.runAsPublic }}
       enablePublishBuildArtifacts: ${{ parameters.enablePublishBuildArtifacts }}
+      
+- ${{ if and(eq(parameters.graphFileGeneration.enabled, true), eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+  - template: ../job/generate-graph-files.yml
+    parameters:
+      continueOnError: ${{ parameters.continueOnError }}
+      includeToolsetInGraph: ${{ parameters.graphFileGeneration.includeToolset }}
+      dependsOn:
+      - ${{ if ne(parameters.publishBuildAssetsDependsOn, '') }}:
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: Wouldn't this just depend on "Asset_Registry_Publish"?

File: eng/common/templates/jobs/jobs.yml [hunk 3]
```diff
@@ -68,4 +74,22 @@ jobs:
         vmImage: vs2017-win2016
       runAsPublic: ${{ parameters.runAsPublic }}
       enablePublishBuildArtifacts: ${{ parameters.enablePublishBuildArtifacts }}
+      
+- ${{ if and(eq(parameters.graphFileGeneration.enabled, true), eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+  - template: ../job/generate-graph-files.yml
+    parameters:
+      continueOnError: ${{ parameters.continueOnError }}
+      includeToolsetInGraph: ${{ parameters.graphFileGeneration.includeToolset }}
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: This needs to be updated

Comment (jcagme) [reply eng/common/templates/jobs/jobs.yml]: Updated how? I am using the new format

File: eng/common/templates/job/generate-graph-files.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+  pool: ${{ parameters.pool }}
+
+  variables:
+  - ${{ if and(eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+    # Publish-Build-Assets provides: MaestroAccessToken, BotAccount-dotnet-maestro-bot-PAT
+    # DotNet-AllOrgs-Darc-Pats provides: dn-bot-devdiv-dnceng-rw-code-pat
+    - group: Publish-Build-Assets
+    - group: DotNet-AllOrgs-Darc-Pats
+    - name: _GraphArguments
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: Note: I haven't tested this at all, but...

With https://github.com/Microsoft/azure-pipelines-yaml/pull/46 now merged, you might be able to write this as something like...

```YAML
- name: _GraphArguments
  value:  -gitHubPat $(BotAccount-dotnet-maestro-bot-PAT)  
        -azdoPat $(dn-bot-devdiv-dnceng-rw-code-pat)  
        -barToken $(MaestroAccessToken)  
        -outputFolder '$(Build.StagingDirectory)/GraphFiles/'
   - ${{ if ne(parameters.includeToolsetInGraph, 'false') }}:
      - name: _GraphArguments
        value: ${{ variables._GraphArguments }} -includeToolset
```

File: eng/common/templates/jobs/jobs.yml (hunk 3 above)
Comment (chcosta) [reply eng/common/templates/jobs/jobs.yml]: Oh, I see.  I expected the `graphFileGeneration` format to flow through.  ie, pass along the `graphFileGeneration` object or name this parameter `includeToolset`.  I'm not certain if it really matters, but at a minimum, I'd name these the same (`includeToolset`)

File: eng/common/templates/jobs/jobs.yml [hunk 5]
```diff
@@ -68,4 +74,16 @@ jobs:
...
+  - template: ../job/generate-graph-files.yml
+    parameters:
+      continueOnError: ${{ parameters.continueOnError }}
+      includeToolset: ${{ parameters.graphFileGeneration.includeToolset }}
+      dependsOn:
+        - Asset_Registry_Publish
+      pool:
+        vmImage: vs2017-win2016
+      runAsPublic: ${{ parameters.runAsPublic }}
+      enablePublishBuildArtifacts: ${{ parameters.enablePublishBuildArtifacts }}
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: This isn't actually used

File: eng/common/templates/job/generate-graph-files.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+parameters:
+  configuration: 'Debug'
+
+  # Optional: condition for the job to run
+  condition: ''
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: This isn't used

File: eng/common/templates/job/generate-graph-files.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
+parameters:
+  configuration: 'Debug'
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: This isn't used

File: eng/common/templates/job/generate-graph-files.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+- job: Generate_Graph_Files
+
+  dependsOn: ${{ parameters.dependsOn }}
+
+  displayName: Generate Graph Files
+
+  pool: ${{ parameters.pool }}
+
+  variables:
+  - ${{ if and(eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: I don't think you need this condition, this thing doesn't run in public, so you can just let it blow up

File: eng/common/templates/job/generate-graph-files.yml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      value: -gitHubPat $(BotAccount-dotnet-maestro-bot-PAT) 
+        -azdoPat $(dn-bot-devdiv-dnceng-rw-code-pat) 
+        -barToken $(MaestroAccessToken) 
+        -outputFolder '$(Build.StagingDirectory)/GraphFiles/'
+    - ${{ if ne(parameters.includeToolset, 'false') }}:
+      - name: _GraphArguments
+        value: ${{ variables._GraphArguments }} -includeToolset
+
+  steps:
+  - ${{ if and(eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: Same as above

File: eng/common/templates/job/generate-graph-files.yml [hunk 10]
```diff
@@ -0,0 +1,54 @@
+parameters:
+  # Optional: dependencies of the job
+  dependsOn: ''
+
+  # Optional: A defined YAML pool - https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=schema#pool
+  pool: {}
+
+  # Optional: should run as a public build even in the internal project
+  #           if 'true', the build won't run any of the internal only steps, even if it is running in non-public projects.
+  runAsPublic: false
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: remove this, let it die if running in public

**Comment by jcagme** [comment]
@chcosta looks good?

File: eng/common/templates/jobs/jobs.yml [hunk 11]
```diff
@@ -68,4 +74,16 @@ jobs:
...
+- ${{ if and(eq(parameters.graphFileGeneration.enabled, true), eq(parameters.runAsPublic, 'false'), ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+  - template: ../job/generate-graph-files.yml
+    parameters:
+      continueOnError: ${{ parameters.continueOnError }}
+      includeToolset: ${{ parameters.graphFileGeneration.includeToolset }}
+      dependsOn:
+        - Asset_Registry_Publish
+      pool:
+        vmImage: vs2017-win2016
+      runAsPublic: ${{ parameters.runAsPublic }}
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: not used

File: eng/common/templates/job/generate-graph-files.yml [hunk 12]
```diff
@@ -0,0 +1,48 @@
+parameters:
+  # Optional: dependencies of the job
+  dependsOn: ''
```
Comment (chcosta) [inline eng/common/templates/job/generate-graph-files.yml]: This should be `dependsOn: []`

File: eng/common/templates/jobs/jobs.yml [hunk 13]
```diff
@@ -68,4 +74,16 @@ jobs:
...
+
```
Comment (chcosta) [inline eng/common/templates/jobs/jobs.yml]: nit: Remove extra line

**Comment by jcagme** [comment]
This build got triggered from here (I think) but checks don't reflect the status. Given that it went fine I'll merge this PR https://dev.azure.com/dnceng/public/_build/results?buildId=121110

**Pull request merged**